doc/developers/contributing.rst

```diff
@@ -207,6 +207,168 @@ and Cython optimizations.
+features of ``rebase``. If you want to squash the latest 3 commits together (this number can
+be found out by typing ``$ git log --oneline``):
+
+	1. Rewind ``HEAD`` to the parent of the last commit you want to edit::
+
+		$ git rebase -i HEAD~3
+
+	2. This will open an editor that might look something like this::
+
+		pick f7f3f6d changed my name a bit
+		pick 310154e updated README formatting and added blame
+		pick a5f4a0d added cat-file
+
+		# Rebase 710f0f8..a5f4a0d onto 710f0f8
+		#
+		# Commands:
+		#  p, pick = use commit
+		#  r, reword = use commit, but edit the commit message
+		#  e, edit = use commit, but stop for amending
+		#  s, squash = use commit, but meld into previous commit
+		#  f, fixup = like "squash", but discard this commit's log message
+		#  x, exec = run command (the rest of the line) using shell
+		#
+		# These lines can be re-ordered; they are executed from top to bottom.
+		#
+		# If you remove a line here THAT COMMIT WILL BE LOST.
+		#
+		# However, if you remove everything, the rebase will be aborted.
+		#
+		# Note that empty commits are commented out
+
+	3. Change ``pick`` to ``fixup`` or ``f`` for the second and third commit like::
+
+		pick f7f3f6d changed my name a bit
+		f 310154e updated README formatting and added blame
+		f a5f4a0d added cat-file
+
+	4. Save and quit the editor. This will immediately popup an editor for editing the
+	   commit message as you're squashing some commits together and might want to change
+	   the original commit message.
+
+	5. After saving and quiting you will see a message like::
+
+		Successfully rebased and updated refs/heads/my-feature-branch.
+
+	6. Once successfully rebased, force push onto your PR branch by::
+
+		$ git push -u -f origin my-feature-branch
+
+
+.. note::
+
+	Detailed guides on **interactive rebasing** can be found:
+
+	* In the *Scipy Development Workflow* in the `rebasing
+	  <http://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#rebasing-on-master>`_
+	  and `squashing
+	  <http://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#
+	  rewriting-commit-history>`_ sections
+
+	* On `gitready.com
+	  <http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html>`_.
+
+	* On `atlassian.com
+	  <https://www.atlassian.com/git/tutorials/merging-vs-rebasing/conceptual-overview>`_.
+
+Rebasing
+^^^^^^^^
+
+You should generally use ``rebase`` over ``merge`` to integrate changes from one branch into
+another branch. Merging creates a "merge commit" every time you incorporate upstream changes.
+This pollutes the feature branch's history and makes it hard for other developers to understand
+the history of the project. On the other hand ``rebase`` effectively rewrites the project history
+by creating brand new commits for each commit in the original branch. This creates a much cleaner
+project history.
+To rebase:
+
+	1. Update your master branch::
+
+		$ git checkout master
+		$ git pull upstream master
```

Comment:
i feel like doing
`git fetch upstream` to get the scikit-learn/scikit-learn changes from master
`git checkout master` to switch to your master branch
`git rebase upstream/master`
is better because git pull creates merge commits, right?